clients/hadoopfs/src/test/java/io/lakefs/LakeFSFileSystemServerTest.java

```diff
@@ -0,0 +1,1375 @@
+        conf.set(org.apache.hadoop.fs.s3a.Constants.SECRET_KEY, S3_SECRET_ACCESS_KEY);
+        conf.set(org.apache.hadoop.fs.s3a.Constants.ENDPOINT, s3Endpoint);
+        conf.set(org.apache.hadoop.fs.s3a.Constants.BUFFER_DIR, "/tmp/s3a");
+
+        conf.set("fs.lakefs.access.key", "unused-but-checked");
+        conf.set("fs.lakefs.secret.key", "unused-but-checked");
+        conf.set("fs.lakefs.endpoint", String.format("http://localhost:%d/", mockServerClient.getPort()));
+        conf.set("fs.lakefs.session_id", sessionId());
+
+        System.setProperty("hadoop.home.dir", "/");
+
+        // lakeFSFS initialization requires a blockstore.
+        mockServerClient.when(request()
+                              .withMethod("GET")
+                              .withPath("/config/storage"),
+                              Times.once())
+            .respond(response()
+                     .withStatusCode(200)
+                     .withBody(gson.toJson(new StorageConfig()
+                                           .blockstoreType("s3")
+                                           .blockstoreNamespaceValidityRegex(".*")
+                                           // TODO(ariels): Change for presigned?
+                                           .preSignSupport(false))));
+
+        // Don't return 404s for unknown paths - they will be emitted for
+        // many bad requests or mocks, and make our life difficult.  Instead
+        // fail using a unique error code.  This has very low priority.
+        mockServerClient.when(request(), Times.unlimited(), TimeToLive.unlimited(), -10000)
+            .respond(response().withStatusCode(418));
+        // TODO(ariels): No tests mock "get underlying filesystem", so this
+        //     also catches its "get repo" call.  Nothing bad happens, but
+        //     this response does show up in logs.
+
+        fs.initialize(new URI("lakefs://repo/main/file.txt"), conf);
+    }
+
+    /**
+     * @return all pathnames under s3Prefix that start with prefix.  (Obvious not scalable!)
+     */
+    protected List<String> getS3FilesByPrefix(String prefix) {
+
+        ListObjectsRequest req = new ListObjectsRequest()
+            .withBucketName(s3Bucket)
+            .withPrefix(prefix)
+            .withDelimiter(null);
+
+        ObjectListing listing = s3Client.listObjects(req);
+        List<S3ObjectSummary> summaries = listing.getObjectSummaries();
+        if (listing.isTruncated()) {
+            Assert.fail(String.format("[internal] no support for test that creates >%d S3 objects", listing.getMaxKeys()));
+        }
+
+        return Lists.transform(summaries, S3ObjectSummary::getKey);
+    }
+
+    @Test
+    public void getUri() {
+        URI u = fs.getUri();
+        Assert.assertNotNull(u);
+    }
+
+    // Expect this statObject to be not found
+    protected void expectStatObjectNotFound(String repo, String ref, String path) {
```

Comment:
Not really an expectation: it's setting the mock to return NotFound. Same for the next ones.

Reply:
I hate the word "mock", but changing words back.